cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java

```diff
@@ -0,0 +1,193 @@
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.http.DefaultHttpRequestConfig;
+import co.cask.cdap.common.internal.remote.MethodArgument;
+import co.cask.cdap.common.internal.remote.RemoteClient;
+import co.cask.cdap.proto.codec.EntityIdTypeAdapter;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.security.spi.authorization.UnauthorizedException;
+import co.cask.common.http.HttpMethod;
+import co.cask.common.http.HttpRequest;
+import co.cask.common.http.HttpResponse;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.Inject;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Remote implementation of the AuthorizationEnforcer. Contacts master for authorization enforcement and
+ * then caches the results if caching is enabled.
+ */
+public class RemoteAuthorizationEnforcer extends AbstractAuthorizationEnforcer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteAuthorizationEnforcer.class);
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
+    .create();
+
+  private final RemoteClient remoteClient;
+  private final boolean cacheEnabled;
+
+  private static LoadingCache<AuthorizationPrivilege, Boolean> authPolicyCache;
+
+  @Inject
+  public RemoteAuthorizationEnforcer(CConfiguration cConf, final DiscoveryServiceClient discoveryClient) {
+    super(cConf);
+    this.remoteClient = new RemoteClient(discoveryClient, Constants.Service.APP_FABRIC_HTTP,
+                                         new DefaultHttpRequestConfig(false), "/v1/execute/");
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    int cacheTtlSecs = cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS);
+    int cacheMaxEntries = cConf.getInt(Constants.Security.Authorization.CACHE_MAX_ENTRIES);
+
+    authPolicyCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<AuthorizationPrivilege, Boolean>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Boolean load(AuthorizationPrivilege authorizationPrivilege) throws Exception {
```

Comment:
So now there is no auto-refresh of cached entries from the container anymore? This means every TTL seconds, the container data operations need to take penalty?